notebooks/en/semantic_cache_chroma_vector_database.ipynb

```diff
@@ -0,0 +1,1389 @@
+{
```

Comment:
<p>In this paragraoh:</p><p> </p><blockquote>The semantic comparison will be performed using the Euclidean distance of question embeddings. This is because, semantically, "What is the capital of France?" is essentially the same as "Tell me the name of the capital of France?"</blockquote><p>I feel like the two sentences do not belong together: the "This is because" does not make sense to me.</p><p>Is it possible to reformulate this? Also, why do you use Euclidian distance rather than the widely used Cosine</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/53/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Reply:
<p>I've modified the introduction:</p><p> </p><p>In this notebook, we will explore a typical RAG solution where we will utilize an open-source model and the vector database Chroma DB.&nbsp;<strong>However, we will integrate a semantic cache system that will store various user queries and decide whether to generate the prompt enriched with information from the vector database or the cache.</strong></p><p> </p><p>A semantic caching system aims to identify similar or identical user requests. When a matching request is found, the system retrieves the corresponding information from the cache, reducing the need to fetch it from the original source.</p><p> </p><p>As the comparison takes into account the semantic meaning of the request, they don't have to be identical for the system to recognize them as the same question. They can be formulated differently or contain inaccuracies, be they typographical or in the sentence structure, and we can identify that the user is actually requesting the same information.</p><p> </p><p>For instance, queries like&nbsp;<strong>What is the capital of France?</strong>,&nbsp;<strong>Tell me the name of the capital of France?</strong>, and&nbsp;<strong>What The capital of France is?</strong>&nbsp;all convey the same intent and should be identified as the same question.</p><p>While the model's response may differ based on the request for a concise answer in the second example, the information retrieved from the vector database should be the same. This is why I'm placing the cache system between the user and the vector database, not between the user and the Large Language Model.</p><p> </p><p>_____</p><p> </p><p>I've used Euclidean distance instead of Cosine because, as far as I know, it's the default distance metric used by Faiss. It can also be calculated with Cosine, but I believe using it introduces added complexity that may not contribute significantly to the explanation.</p><p><a href="https://github.com/facebookresearch/faiss/issues/95" target="_blank">https://github.com/facebookresearch/faiss/issues/95</a></p><p> </p><p> </p>

Reply:
<p>I think this justification about why to use the Euclidean distance is perfectly sound, you should put it somewhere in the notebook!</p>

Reply:
<p>Hi, what about this text: </p><p> </p><p>I've used Euclidean distance instead of Cosine, which is widely employed in vector comparisons. This choice is based on the fact that Euclidean distance is the default metric used by Faiss. Although Cosine distance can also be calculated, doing so adds complexity that may not significantly contribute to the final result.</p><p> </p><p> </p>